### ISSUE TITLE
nginx http redirect

### ISSUE THREAD

**Issue by vasilevich** [certbot/certbot#1825]
Apologize in advance, I would edit it myself but lack the knowledge in python.

the HTTP redirect is implemented wrong:
example of current implementation

SERVER{
if(its http)
   {redirect to HTTP}
 listen 443 ssl;
}
problem: http traffic will never reach this block due to listen 443.
instead you should make an entirely new block
with the exact same server names:
server{
listen 80; #is default anyway
   server names exactly like in the edited block...
return ...what's in the if;
}

**Comment by vasilevich** [comment]
apologizes again forgot to add, the issuse should be at the file 

letsencrypt-nginx/letsencrypt_nginx/configurator.py

lines: 359-388

**Comment by Queeq** [comment]
AFAIK the best practice is this:

```
server {
        listen *:80;
        server_name example.com;
        # Redirect to https
        return 301 https://$server_name$request_uri;
}
```
`listen` and `server_name` should be taken from original configuration, definitely.

**Comment by vasilevich** [comment]
I added this issuse from my phone.
here are some more details and the procedure to reproduce the error:
Nginx settings before running the script:
server {
    include /etc/nginx/subextensions/*;
    server_name www.duckgamer.com;
    server_name duckgamer.com;
    root $Websites/duckgamer;

    location / {

        }
}

, after running the script(in arch linux with default setup of pacman -S letsencrypt letsencrypt-nginx):
letsencrypt -d duckgamer.com -d www.duckgamer.com --standalone
in the easy/secure window, selected:
secure make all requests reditect to secure HTTPS access
new nginx config:
server {

    if ($scheme != "https") {
        return 301 https://$host$request_uri;
    }
    ssl_stapling_verify on;
    ssl_stapling on;
    ssl_trusted_certificate /etc/letsencrypt/live/duckgamer.com/chain.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_certificate_key /etc/letsencrypt/live/duckgamer.com/privkey.pem;
    ssl_certificate /etc/letsencrypt/live/duckgamer.com/fullchain.pem;
    error_log /var/lib/letsencrypt/error.log;
    access_log /var/lib/letsencrypt/access.log;
    listen 443 ssl;
    include /etc/nginx/subextensions/*;
    server_name www.duckgamer.com;
    server_name duckgamer.com;
    root $Websites/duckgamer;

    location / {
    }
}

The problem in this config is the combination of:
 if ($scheme != "https") {
        return 301 https://$host$request_uri;
    }

with:
 listen 443 ssl;
HTTP requests never reach this server block because of the listen 443 ssl.
rendering the HTTP protocol entirely inaccesible.

suggested fix:
Instead of putting
if ($scheme != "https") {
        return 301 https://$host$request_uri;
    }
inside the same block
you should put the whole redirect in its own SERVER block here is my approach:

server {
    ssl_stapling_verify on;
    ssl_stapling on;
    ssl_trusted_certificate /etc/letsencrypt/live/duckgamer.com/chain.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_certificate_key /etc/letsencrypt/live/duckgamer.com/privkey.pem;
    ssl_certificate /etc/letsencrypt/live/duckgamer.com/fullchain.pem;
    error_log /var/lib/letsencrypt/error.log;
    access_log /var/lib/letsencrypt/access.log;
    listen 443 ssl;
    include /etc/nginx/subextensions/*;
    server_name www.duckgamer.com;
    server_name duckgamer.com;
    root $Websites/duckgamer;
    location / {
    }
}

server{
    server_name www.duckgamer.com;
    server_name duckgamer.com;
  return 301 https://$host$request_uri;
}

this is both, a working + the recommended approach since "ifs" are evil by the NGINX recommendations.
https://www.nginx.com/resources/wiki/start/topics/depth/ifisevil/

Thank you for your time.

**Comment by crispgm** [comment]
You may try this way,

```
server {
    listen 80;
    # other stuffs like server_name
    location / {
        rewrite "^/(.*)$" https://$host/$1 break;
    }
}
```

Kind regards.

**Comment by vasilevich** [comment]
yes but that approach is good if you want to make ALL your domains strict https which is not always the case, my approach is more friendly for multi protocol configs, moreover I think a regex-less approach is a tiny bit better in terms of perfomance and if you remove the server_name you should get a similar result as yours I suppose.

**Comment by bmw** [comment]
To copy over my comment from #3351, we may want to remove redirect as a supported enhancement temporarily in Nginx for the 0.9.0 release.

**Comment by bmw** [comment]
#3505 is related for how we plan to enable the redirect

**Comment by bmw** [comment]
Fixed in #3546.
